### PR TITLE
Manage Fortios/Fortigate Address

### PR DESCRIPTION
**Pull request by bjolivot**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

fortios_address

##### ANSIBLE VERSION

```
ansible 2.4.0 (fortios_address 3b641216d6) last updated 2017/05/15 16:49:14 (GMT +200)
  config file = /home/ben/.ansible.cfg
  configured module search path = [u'/home/ben/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ben/ansible/ansible_src/lib/ansible
  executable location = /home/ben/ansible/ansible_src/bin/ansible
  python version = 2.7.9 (default, Jun 29 2016, 13:08:31) [GCC 4.9.2]
```

##### SUMMARY

New fortios module

### CODE REVIEW

**Comment by bjolivot** [comment]
cc @resmo I had to close the other PR as I messed up with rebase

File: lib/ansible/modules/network/fortios/fortios_address_group.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+firewall_address_config:
+  description: full firewall adresses config string
+  returned: always
+  type: string
+change_string:
+  description: The commands executed by the module
+  returned: only if config changed
+  type: string
+"""
+import logging
```
Comment (resmo) [inline lib/ansible/modules/network/fortios/fortios_address_group.py]: This seems to be a left over

File: lib/ansible/modules/network/fortios/fortios_address_group.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+def main():
+    #define module params
+    argument_spec = dict(
+        host        = dict(required=True ),
+        username    = dict(required=True ),
+        password    = dict(required=True, type='str', no_log=True ),
+        timeout     = dict(type='int', default=60),
+        vdom        = dict(type='str' ),
+        state       = dict(required=True, choices=['present', 'absent']),
```
Comment (resmo) [inline lib/ansible/modules/network/fortios/fortios_address_group.py]: in many modules, `state=present` is the default. If it makes sense to you, consider it too.

File: lib/ansible/modules/network/fortios/fortios_ipv4_policy.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    default: any
+  dstintf:
+    description:
+      - Specifies destination interface name.
+    default: any
+  srcaddr:
+    description:
+      - Specifies source address (or group) object name.
+    required: true
+  srcaddr-negate:
```
Comment (resmo) [inline lib/ansible/modules/network/fortios/fortios_ipv4_policy.py]: In the arg spec, this is written like:  `srcaddr_negate`, for better readability, I would even change into: `src_addr_negate` but this is up to you.

File: lib/ansible/modules/network/fortios/fortios_ipv4_policy.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+  action:
+    description:
+      - Specifies accept or deny action policy.
+    choices: ['accept', 'deny']
+    required: true
+  service:
+    description:
+      - "Specifies policy service(s), could be a list (ex: ['MAIL','DNS'])."
+    required: true
+  service-negate:
```
Comment (resmo) [inline lib/ansible/modules/network/fortios/fortios_ipv4_policy.py]: change to `service_negate` according arg_spec

File: lib/ansible/modules/network/fortios/fortios_ipv4_policy.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    description:
+      - Negate policy service(s) defined in service value.
+    choices: ['enable','disable']
```
Comment (resmo) [inline lib/ansible/modules/network/fortios/fortios_ipv4_policy.py]: as ansible is about "declarative", it should rather be called like `enabled` than `enable` e.g. `choices: ['enabled', 'disabled']`. 

Optionally use booleans for it as it is on/off.  `service_negate_enabled: true` or `service_negate: true`. But this is up to you.

File: lib/ansible/modules/network/fortios/fortios_ipv4_policy.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+        id            = dict(type='str', required=True),
+        srcintf       = dict(default='any'),
+        dstintf       = dict(default='any'),
+        state         = dict(required=True, choices=['present', 'absent']),
+        srcaddr       = dict(required=True, type='str'),
+        dstaddr       = dict(required=True, type='str'),
+        srcaddr_negate= dict(choices=['enable', 'disable'], default='disable'),
+        dstaddr_negate= dict(choices=['enable', 'disable'], default='disable'),
+        action        = dict(choices=['accept','deny'], required=True),
+        service       = dict(required=True, type='list'),
```
Comment (resmo) [inline lib/ansible/modules/network/fortios/fortios_ipv4_policy.py]: as this is a list, also add an alias `aliases=['services']`

~~~
- name: Allow external DNS call
  fortios_ipv4_policy:
    services: 
    - dns
    - ntp
    - http
...
~~~

**Review by resmo — COMMENTED** [review]
needs_revision

**Comment by mattclay** [comment]
CI failure due to module docs validation error:

```
2017-02-21 21:58:58 ============================================================================
2017-02-21 21:58:58 lib/ansible/modules/network/fortios/fortios_ipv4_policy.py
2017-02-21 21:58:58 ============================================================================
2017-02-21 21:58:58 ERROR:305:DOCUMENTATION.options.service.aliases: expected list for dictionary value @ data['options']['service']['aliases']. Got None
```

**Comment by bjolivot** [comment]
As best practices defined here:
http://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html
this PR will be splitted in 3, this one only for fortios_address module

**Comment by bjolivot** [comment]
ready_for_review

**Comment by bjolivot** [comment]
Wait before review I will release common code in module_utils

**Comment by calfonso** [comment]
@bjolivot just checking in here, are you planning on continuing work on module_utils, and what would you like to do with this PR?

**Comment by bjolivot** [comment]
@calfonso  I was waiting for #23275 to be merged before working on this PR

**Comment by bjolivot** [comment]
-label WIP 
-label affects_2.3

**Comment by bjolivot** [comment]
ready_for_review

**Comment by gundalow** [comment]
Closing/reopening to trigger CI as we've made a number of improvements recently

**Pull request merged**